projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/SupportedPythonVersions.java

```diff
@@ -0,0 +1,130 @@
+   * 3.8, ...]. If the supportedPythonVersions configuration is not set, the method returns the
+   * default python version set in the deploymentDataJobBaseImage configuration property.
+   *
+   * @return A string of all python versions supported by the Control Service
+   */
+  public String getSupportedPythonVersions() {
+    if (!supportedPythonVersions.isEmpty()) {
+      return supportedPythonVersions.keySet().toString();
+    } else {
+      try {
+        return "[" + getDefaultPythonVersion() + "]";
+      } catch (IllegalArgumentException e) {
+        throw new IllegalArgumentException(e);
+      }
+    }
+  }
+
+  /**
+   * Returns the name of the data job base image as stored in the docker registry. If
+   * supportedPythonVersions is set, and the python_version passed by the user is supported
+   * according to the configuration, the base image corresponding to the python_version is returned.
+   * Otherwise, the default base image name as set in deploymentDataJobBaseImage is returned.
+   *
+   * @param python_version a string indicating the python version passed by the user
+   * @return a string of the data job base image.
+   */
+  public String getJobBaseImage(String python_version) {
+    if (!supportedPythonVersions.isEmpty() && isPythonVersionSupported(python_version)) {
+      return supportedPythonVersions.get(python_version).get(BASE_IMAGE);
+    } else {
+      return deploymentDataJobBaseImage;
+    }
+  }
+
+  public String getDefaultJobBaseImage() {
+    return deploymentDataJobBaseImage;
+  }
+
+  /**
+   * Returns the name of the vdk image as stored in the docker registry. If supportedPythonVersions
+   * is set, and the python_version, passed by the user, is supported according to the
+   * configuration, the vdk image corresponding to the python_version is returned. Otherwise, the
+   * default vdk image name as set in vdkImage is returned.
+   *
+   * @param python_version a string indicating the python version passed by the user
+   * @return a string of the data job base image.
+   */
+  public String getVdkImage(String python_version) {
+    if (!supportedPythonVersions.isEmpty() && isPythonVersionSupported(python_version)) {
+      return supportedPythonVersions.get(python_version).get(VDK_IMAGE);
+    } else {
+      return vdkImage;
+    }
+  }
+
+  public String getDefaultVdkImage() {
+    return vdkImage;
+  }
+
+  /**
+   * Returns the default python version supported by the Control Service. The version number is
+   * extracted from the datajobs.deployment.dataJobBaseImage application property. The property is
+   * set as for example, `python:3.9-slim`, and we use a regex to match `3.9` and return it to the
+   * caller.
+   *
+   * @return a string indicating the default python version supported by the Control Service.
+   * @throws IllegalArgumentException
+   */
+  public String getDefaultPythonVersion() throws IllegalArgumentException {
```

Comment:
I don't think it is a good approach, definitely won't work in all cases.

Reply:
Can you elaborate? The base images we build contain the python version number, and unless we introduce yet another application property, we don't have any way of knowing what version is the default.

Reply:
Such logic is error-prone if the administrator passes the image without the python version in it.

Reply:
I am in favour of `datajobs.deployment.defaultPythonVersion` property and deprecating those two. 

Overall we add one and remove 2 so we are one ahead :)

If user do not specify datajobs.deployment.defaultPythonVersion then we could default to latest supported version. 
If we specify default version that is not supported we could default to closest one (and generate error message)

I think this is better because 
A) As [Zen of Python](https://peps.python.org/pep-0020/) says "explicit is better than implicit"
B) this it would be easier to maintain and update, as the Python version can be changed independently of the base image name 
C) Code is clearer and less error-prone (one advice from me - always be wary of introducing parsing logic unless you follow well-defined standards (RFC / IEEE level)  and it guarantees that something you need will be there in the format.)

Reply:
@tozka @mivanov1988 

Did a complete refactoring of the logic and removed the use of the `datajobs.vdk.image` and `datajobs.deployment.dataJobBaseImage` application properties, and replaced them with `datajobs.deployment.supportedPythonVersions` and `datajobs.deployment.defaultPythonVersion` properties. Could you, please, do another pass of the PR? Thank you